### ISSUE TITLE
Casting loads from bool tensors to tl.int1 causes hangs on ROCm

### ISSUE THREAD

**Issue by jataylo** [ROCm/triton#353]
@jayfurmanek @zhanglx13 @binarman
cc: @dllehr-amd @jeffdaily 

A PyTorch PR was merged which casts triton loads from boolean tensors to `tl.int1` https://github.com/pytorch/pytorch/pull/110388. This change causes ROCm to hang even on triton-mlir TOT.

Reproducing triton code
```
import torch
import math
import random
import os

import triton
import triton.language as tl
from torch._dynamo.testing import rand_strided
from torch._C import _cuda_getCurrentRawStream as get_cuda_stream
from torch import empty_strided

@triton.jit
def triton_kn(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 20
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), xmask).to(tl.int1)
    tmp1 = tmp0.to(tl.float32)
    tmp2 = tl.math.exp2(tmp1)
    tl.store(out_ptr0 + (x0), tmp2, xmask)

arg0_1 = rand_strided((20, ), (1, ), device='cuda:0', dtype=torch.bool)
buf0 = empty_strided((20, ), (1, ), device='cuda:0', dtype=torch.float32)
stream0 = get_cuda_stream(0)
triton_kn[(1,)](arg0_1, buf0, 1, 1, stream=stream0)
```

The same code passes if the cast is removed:
`tmp0 = tl.load(in_ptr0 + (x0), xmask)`

For now we have reverted this change for ROCm on PyTorch.

**Comment by oplavsic** [comment]
## Investigation
### Difference between hanging and non-hanging kernels

As @zhanglx13 noticed in the previous comment, there is difference in function calls generated for hanging and non-hanging kernels. In case of a hanging kernel, exp2 function call is replaced by ldexpf function call. This is LLVM middle end (target independent) optimization implemented in llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp file in function LibCallSimplifier::optimizeExp2. Goal of this function is to perform transforms:
  1) exp2(sitofp(x)) -> ldexp(1.0, sext(x))  if sizeof(x) <= IntSize
  2) exp2(uitofp(x)) -> ldexp(1.0, zext(x))  if sizeof(x) < IntSize

In general ldexp is simpler and more performant than exp2 since is requires only simple bit manipulation (source:https://stackoverflow.com/questions/39587752/difference-between-ldexp1-x-and-exp2x). Additionally, conversion from int to float is not required which saves few cycles as well.

###why is this optimization not performed in 'non-hanging' case?
In non hanging case, we use int8 which is a signed type, so situation 1) occurs. However, in signed type case, there are some previous transformations that ensure that subnormal values are handled properly, so IR before optimizeExp2 optimization looks like this:
  %14 = load i8, ptr %13, align 1, !dbg !14
  %15 = sitofp i8 %14 to float, !dbg !17
  %16 = icmp slt i8 %14, -126, !dbg !18
  %17 = select i1 %16, float 0x3BF0000000000000, float 1.000000e+00, !dbg !18
  %18 = select i1 %16, float 6.400000e+01, float 0.000000e+00, !dbg !18
  %19 = fadd float %18, %15, !dbg !18
  %20 = tail call float @llvm.exp2.f32(float %19), !dbg !18

This won't be pattern matched, because optimizeExp2 only pattern matches code like this:
  %19 = sitofp i8 %14 to float, !dbg !17
  %20 = tail call float @llvm.exp2.f32(float %19), !dbg !18.

In int1 case, we have unsigned type, so there is no handling of subnormal values, so IR actually looks something like:
  %19 = uitofp i8 %14 to float, !dbg !17
  %20 = tail call float @llvm.exp2.f32(float %19), !dbg !18
Which is pattern matched in optimizeExp2 and thus optimization happens replacing exp2 with ldexp.

### Why doesn't ldexp work properly?

As @zhanglx13 mentioned, in assembly we see call to library function ldexpf@gotpcrel32@lo+4. This is not supposed to happen. Rocm math libraries are shipped as bitcodes, and are supposed to be linked and inlined during compilation.
I was looking trough the llvm backend code  to see how is ldexp handled, and from what I could see, it appears that there is no proper handling of ldexp call in llvm 17 that we are currently using. I found newer commit in LLVM backend related to this: https://github.com/llvm/llvm-project/commit/0eabe65bfbb8eed5d0ffb3718fed3c9bbe058379.
Currently triton doesn't build with this version of llvm (where this commit is introduced) so I can't actually test if this fixes things, but it looks promising. 
Other other thing that it could be, is that ocml library that we are using and llvm version are not in sync so different ldexp mnemonic is used, thus llvm is not able to reason about it. I will try to check if this is the case.

I think we should probably wait until next IFU lands, and we are able to build triton with llvm that uses this new handling of ldexp. If we still see the problem, we can debug further or even create ticket for LLVM backend guys.

To sum up, this looks like a backend issue to me, where ldexp is not properly lowered to llvm intrinsics.

**Comment by jerryyin** [comment]
Closing as completed
